mailpoet/lib/Automation/Engine/Storage/WorkflowStorage.php

```diff
@@ -131,6 +142,39 @@ public function getActiveWorkflowsByTrigger(Trigger $trigger): array {
     }, (array)$data);
   }
 
+  public function deleteWorkflow(Workflow $workflow): void {
```

Comment:
I struggled with where this logic should live. On the one hand it doesn't feel right to have `WorkflowStorage` handle deleting runs and logs, but it also didn't seem appropriate to only have it take place in the API endpoint, because in an ideal world I think we would handle this type of thing directly at the database level by using foreign key constraints with cascading deletes, but since some of our users use MyISAM (and probably for other reasons) that's not a viable option.

I'm totally open to a different approach though! I thought about using some actions, like `beforeDeleteWorkflow` or something that would allow other parts of the application to respond to deletion events, but I wasn't sure where we would actually put the logic for deleting related records.

Reply:
I think it's fine, what can we do... We could inject other storages and reuse them but I'm not sure if a storage should reuse another one. We may end up with cyclic references.